src/azul/chalice.py

```diff
@@ -504,34 +498,108 @@ def swagger_resource(self, file) -> Response:
+            lambda_name, _ = config.unqualified_resource_name(app_name)
+            return attr.evolve(self,
+                               lambda_name=lambda_name,
+                               handler_name=handler_name)
+
+        @property
+        def lambda_tf_resource_name(self) -> str:
+            if self.handler_name is None:
+                return self.lambda_name
+            else:
+                assert self.handler_name != ''
+                return f'{self.lambda_name}_{self.handler_name}'
+
+    # noinspection PyPep8Naming
+    @attr.define(frozen=True, kw_only=True)
+    class metric_alarm(lambda_decorator):
+        """
+        Use this decorator to configure a metric alarm for a Lambda function.
+        This decorator cannot be used to decorate view functions, i.e. functions
+        also decorated with ``@app.route``.
+        """
+        #: The AWS/Lambdas namespace for this alarm
+        metric: LambdaMetric
+        #: The number of failed or throttled lambda invocations that, when
+        #: exceeded, will trigger the alarm.
+        threshold: int = attr.ib(default=0)
+        #: The interval (in seconds) at which the alarm threshold is evaluated,
+        #: ranging from 1 minute to 1 day. The default is 5 minutes.
+        period: int = attr.ib(default=5 * 60)
+
+        def __call__(self, f):
             assert isinstance(f, chalice.app.EventSourceHandler), f
-            f.errors_threshold = errors
-            f.throttles_threshold = throttles
+            try:
+                metric_alarms = getattr(f, 'metric_alarms')
+            except AttributeError:
+                metric_alarms = f.metric_alarms = []
+            metric_alarms.append(self)
             return f
 
-        return wrapper
+        @property
+        def alarm_tf_resource_name(self) -> str:
+            return f'{self.lambda_tf_resource_name}_{self.metric.name}'
+
+    # noinspection PyPep8Naming
+    @attr.define(frozen=True)
+    class retry(lambda_decorator):
+        """
+        Use this decorator to specify the number of times a Lambda invocation of
+        the decorated event handler function should be retried. This decorator
+        cannot be used to decorate view functions, i.e. functions also decorated
+        with ``@app.route``.
+
+        https://docs.aws.amazon.com/lambda/latest/dg/invocation-retries.html
+        """
+        times: int
+
+        def __call__(self, f):
+            assert isinstance(f, chalice.app.EventSourceHandler), f
+            f.retry = self
+            return f
 
     @property
-    def metric_thresholds(self) -> list[MetricThreshold]:
-        default_threshold = 0
-        thresholds = []
-        lambda_name, _ = config.unqualified_resource_name(self.app_name)
+    def retries(self) -> Iterator[retry]:
+        for handler_name, handler in self.handler_map.items():
+            if not isinstance(handler, chalice.app.EventSourceHandler):
+                pass
+            else:
```

Comment:
I may have given conflicting feedback in an earlier review. I apologize for that, but having observed some recent reviews by Hannes I now believe the `pass` branch should be eliminated.

```suggestion
            if isinstance(handler, chalice.app.EventSourceHandler):
```